### PR TITLE
Avoid restarting the TestingSiloHost

### PR DESCRIPTION
**Pull request by jdom**
Previously it was restarting the silos every time a test started. Now we try to preserve the previous silo host whenever we can.

This is WIP, not that far from finishing, but I wanted to show a little progress before polishing it, in case someone was interested. I will wait for #1255 to be merged and then rebase on top of that one after fixing the issues that appear from those new tests (as I changed the UnitTestsSiloHost constructor)

### CODE REVIEW

File: src/OrleansTestingHost/TestingSiloHost.cs [hunk 1]
```diff
@@ -671,7 +738,7 @@ private static void StopOrleansSilo(SiloHandle instance, bool stopGracefully)
             catch (Exception exc)
             {
                 Console.WriteLine(exc);
-                throw;
+                //throw;
```
Comment (jthelin) [inline src/OrleansTestingHost/TestingSiloHost.cs]: Might want to add to the WriteLine above something along the lines of "While stopping silo {instance} ignoring error {exception}" to make it clearer [a] what that console message relates to, [b] that the error is being treated as non-fatal, and [c] which silo instance caused the problem.

File: src/Tester/BasicActivationTests.cs [hunk 2]
```diff
@@ -15,36 +14,26 @@
...
         {
-            ITestGrain g1 = GrainClient.GrainFactory.GetGrain<ITestGrain>(1);
-            ITestGrain g2 = GrainClient.GrainFactory.GetGrain<ITestGrain>(2);
-            Assert.AreEqual(1L, g1.GetPrimaryKeyLong());
-            Assert.AreEqual(1L, g1.GetKey().Result);
-            Assert.AreEqual("1", g1.GetLabel().Result);
-            Assert.AreEqual(2L, g2.GetKey().Result);
-            Assert.AreEqual("2", g2.GetLabel().Result);
+            long g1Key = GetRandomGrainId();
+            long g2Key = GetRandomGrainId();
```
Comment (jthelin) [inline src/Tester/BasicActivationTests.cs]: Some parts of this change set like this could be split into separate PR because I don't think they have a direct dependency on the no-restart part of the change? 
This test should work even under the current TestingSiloHost implementation, won't it?

File: src/Tester/FSharpGrainTests.cs [hunk 3]
```diff
@@ -14,20 +12,9 @@ namespace UnitTests.General
     /// Unit tests for grains implementing generic interfaces
     /// </summary>
     [TestClass]
-    public class FSharpGrainTests : UnitTestSiloHost
+    public class FSharpGrainTests : UnitTestSiloHostEnsureDefaultStartedSecondaryOff
```
Comment (jthelin) [inline src/Tester/FSharpGrainTests.cs]: Would it be clearer if this was renamed to `UnitTestSiloHostEnsureSingleSiloStarted` ?

File: src/Tester/StreamingTests/PullingAgentManagementTests.cs [hunk 4]
```diff
@@ -16,30 +16,23 @@ namespace UnitTests.StreamingTests
     [DeploymentItem("ClientConfigurationForStreamTesting.xml")]
     [DeploymentItem("OrleansProviders.dll")]
     [TestClass]
-    public class PullingAgentManagementTests : UnitTestSiloHost
+    public class PullingAgentManagementTests : UnitTestSiloHostPerFixture
```
Comment (jthelin) [inline src/Tester/StreamingTests/PullingAgentManagementTests.cs]: Does the word `Fixture` here have the same semantic meaning as it does in the xUnit test framework?
http://xunit.github.io/docs/shared-context.html

File: src/Tester/UnitTestSiloHost.cs [hunk 5]
```diff
@@ -85,7 +79,7 @@ public static string DumpTestContext(TestContext context)
             return sb.ToString();
         }
 
-        public static int GetRandomGrainId()
+        public static long GetRandomGrainId()
```
Comment (jthelin) [inline src/Tester/UnitTestSiloHost.cs]: Good catch!

File: src/Tester/UnitTestSiloHost.cs [hunk 6]
```diff
@@ -25,24 +28,13 @@ namespace UnitTests.Tester
...
     {
-        public UnitTestSiloHost() // : base()
-        {
-        }
-
-        public UnitTestSiloHost(bool startFreshOrleans)
-            : base(startFreshOrleans)
-        {
-        }
+        protected static readonly Random random = new Random();
```
Comment (jthelin) [inline src/Tester/UnitTestSiloHost.cs]: Probably better to use `SafeRandom` here, just to ensure there are no unforeseen glitches if any test cases try to use random inside concurrent executing closures etc.

File: src/TesterInternal/TimerTests/ReminderTests_AzureTable.cs [hunk 7]
```diff
@@ -300,12 +287,12 @@ public void TestCleanup()
             Task<bool> t1 = Task.Factory.StartNew(() =>
             {
 
-                StopSilo(siloToKill);
+                siloHost.StopSilo(siloToKill);
```
Comment (jthelin) [inline src/TesterInternal/TimerTests/ReminderTests_AzureTable.cs]: Every time i see the `siloHost` i get confused about what it is!
1) it seems like it should have a capital 'S' to stop it looking so much like a private field in this class - it's a "property" shared / inherited from the base class, isn't it? From the name, i don't know where to expect it to "live".
2) isn't that object a "handle" wrapping an in-process mini-cluster of multiple silos, rather than just a single "silo"? Would a name along the lines of `HostedTestCluster.StopSilo(x)` make that clearer?
3) There seems to be a natural evolution from this "single cluster" to the "multi cluster" tests what @sebastianburckhardt is working on. Will we re-use this "cluster host" directly for the multi-cluster cases?

**Comment by jthelin** [comment]
I added a few specific feedback comments above, but nothing fundamental or serious.

Seems like a very good move to clean up lots of this stuff up. Nice job!

File: src/OrleansTestingHost/TestingSiloHost.cs (hunk 1 above)
Comment (jdom) [inline src/OrleansTestingHost/TestingSiloHost.cs]: hmmm, I meant to revert this change

File: src/Tester/BasicActivationTests.cs (hunk 2 above)
Comment (jdom) [inline src/Tester/BasicActivationTests.cs]: yes, this test "should" work in both environments, as it's an improved version of the tests. Nevertheless it's more important now with the changes, as using fixed IDs is a bad idea if you don't restart the silos (there were a few tests trying to use the same grains and affecting each other). So even if I could improve the tests in a separate PR, it's something that we probably wouldn't just do if this other change wasn't coming too.

File: src/Tester/FSharpGrainTests.cs (hunk 3 above)
Comment (jdom) [inline src/Tester/FSharpGrainTests.cs]: I'm all up for a better name. Nevertheless, not sure if "Single" is the right word. I tried to imply that I started the silo host with the default configuration (with no providers that depend on Azure, no weird config settings, etc). Currently it's not even a single silo, but 2 (although that might change if I see that it speeds up things significantly).
Can you recommend a better name, given the explanation?

File: src/Tester/StreamingTests/PullingAgentManagementTests.cs (hunk 4 above)
Comment (jdom) [inline src/Tester/StreamingTests/PullingAgentManagementTests.cs]: well, today it would map to a class fixture. Although after the migration I intend to map it to the broader collection fixture, and with that have even less restarts.

File: src/Tester/UnitTestSiloHost.cs (hunk 6 above)
Comment (jdom) [inline src/Tester/UnitTestSiloHost.cs]: yes, I could copy that class to this project (or link it, as it's an internal that I don't have access too). I thought about doing that. I wouldn't want to make that utility class public as it has nothing to do with Orleans.

File: src/TesterInternal/TimerTests/ReminderTests_AzureTable.cs (hunk 7 above)
Comment (jdom) [inline src/TesterInternal/TimerTests/ReminderTests_AzureTable.cs]: 1) yes, making this a property and pascal casing it was one of the "polishing" tasks that I had in my TODO list.
2) I like the idea of calling it something else, I just used SiloHost as a shorthand for TestingSiloHost which is the current class name we are using. I like HostedCluster better as a property name (but are you suggesting also renaming the class? that's a bigger breaking change, not sure how many folks are using it). Nevertheless, it's technically more than that, as it abstract both the cluster and the client in it.
3) Sure, I think it can evolve to account for both:
 - multi-cluster: although not sure how many tests will need that, apart from the tests that actually assert that the multi-cluster feature works. I mean, I don't see to many Orleans consumers using it, but I might be wrong.
 - Client isolated in it's app-domain: this could be very useful to be able to parallelize many fixtures, as they could all start their testing cluster with randomized ports, and also their clients in a separate app domain. This work is not trivial (Sebastian is doing it, but it requires some massaging in the way you write the tests), so that's why I didn't do it as part of this PR. I definitely would like to see it evolve in that direction, and rely less and less on statics. Even better, if I could make Orleans itself not rely on statics, even better, no need for a separate AppDomain for the clients

File: src/Tester/FSharpGrainTests.cs (hunk 3 above)
Comment (jthelin) [inline src/Tester/FSharpGrainTests.cs]: But the original version of this test particular was only using a single silo (Primary, but no Secondary)

``` c#
public FSharpGrainTests()
    : base(new TestingSiloOptions { StartPrimary = true, StartSecondary = false })
{
}
 ```

Isn't that why you used the `SecondaryOff` part of the name `UnitTestSiloHostEnsureDefaultStartedSecondaryOff` ?

I think the `UnitTestSiloHostEnsureDefaultStarted` name for the "normal" (2-silo) case is ok, 
but `UnitTestSiloHostEnsureDefaultStartedSecondaryOff` seems to bury the headline that this test case is only using single-silo.
But i do take your point that the other part of the name is the "Defaults" part to signify using the baseline "default" config.

How about something link this:
`UnitTestSiloHostEnsureDefaultSingleSiloStarted`

BTW, i don't think we should get too hung up on finding the perfect name here, so this is P2 issues in my book - way more important to get this change set merged than to rat hole on names.

Comment (jthelin) [inline src/Tester/FSharpGrainTests.cs]: I guess a higher level question is whether these test cases should even be using a special case config of single silo, rather than using the same "normal" test config that all other tests uses?

File: src/Tester/StreamingTests/PullingAgentManagementTests.cs (hunk 4 above)
Comment (jthelin) [inline src/Tester/StreamingTests/PullingAgentManagementTests.cs]: Cool. I have played around with xUnit on another project, and after factoring out the shared context into "fixtures" then porting from MsTest to xUnit is very straight forward task :)

File: src/TesterInternal/TimerTests/ReminderTests_AzureTable.cs (hunk 7 above)
Comment (jthelin) [inline src/TesterInternal/TimerTests/ReminderTests_AzureTable.cs]: > I like HostedCluster better as a property name (but are you suggesting also renaming the class? ...)

No, just the variable / property usage in these test suites.

File: src/Tester/FSharpGrainTests.cs (hunk 3 above)
Comment (jdom) [inline src/Tester/FSharpGrainTests.cs]: ohh, right, I missed that... yeah, all of these marker classes I left there (UnitTestSiloHostEnsureDefaultStarted2, UnitTestSiloHostEnsureDefaultStarted3, UnitTestSiloHostEnsureDefaultStartedSecondaryOff) were left there for me as I was doing the changes, to identify issues if tests failed, to know that the original config was different. Nevertheless, that is the main thing I wanted to polish, which is remove all the unnecessary different configs... So as you also suggested, I don't plan to have different cluster configs between these once I submit the polished branch, and I'll move the remaining base classes to their own code files

File: src/Tester/StreamingTests/PullingAgentManagementTests.cs (hunk 4 above)
Comment (jdom) [inline src/Tester/StreamingTests/PullingAgentManagementTests.cs]: yeah, part of this refactoring was also trying to avoid relying "too much" on MSTest-specific features, so that I don't complicate the upcoming migration unnecessarily

**Comment by jdom** [comment]
Thanks @jthelin for all the feedback. Btw, I worked a little bit more, as well as rebased and fixed after the last vso tests migration. Tomorrow I'll do some more minor cleanup and should be ready for merge

**Comment by jdom** [comment]
Addressed all the feedback and rebased. I don't have any more work planned, so it should be ready to merge if nobody else has objections

File: src/TesterInternal/StreamingTests/SMSStreamingTests.cs [hunk 8]
```diff
@@ -35,30 +35,23 @@ public class SMSStreamingTests : UnitTestSiloHost
...
-            runner = new SingleStreamTestRunner(SingleStreamTestRunner.SMS_STREAM_PROVIDER_NAME);
-            fireAndForgetDeliveryProperty = ExtractFireAndForgetDeliveryProperty();
+            return new TestingSiloHost(
+                smsSiloOption,
+                //smsSiloOption_OnlyPrimary,
+                smsClientOptions);
         }
 
-        public SMSStreamingTests(int dummy)
-            : base(smsSiloOption_OnlyPrimary, smsClientOptions)
```
Comment (jdom) [inline src/TesterInternal/StreamingTests/SMSStreamingTests.cs]: there were some tests like this that had a non-default constructor, but no derived class, so not sure what was the purpose as MSTest would never run those tests. I kept some code commented out (such as the `//smsSiloOption_OnlyPrimary` in the CreateSiloHost method) but not sure about the usefulness of it. @sergeybykov  do you have any recommendation? If those tests are valid, then I could create a derived class that exercises that path.

Comment (jdom) [inline src/TesterInternal/StreamingTests/SMSStreamingTests.cs]: or maybe @gabikliot knows if we care about these non-default constructors... I think it's not worth spending any time on those, just highlighting that with this changes, the context of the old constructors might be lost, in case we cared about them. I suppose they were just some dev-only tests as the author was troubleshooting issues
